File: src/mainview/features/installs.ts
import type { PluginDefinitionInfo } from "../../shared/contracts";
import { rpc } from "../core/rpc";
import { state, setBusy } from "../core/state.svelte";
import type { MainViewState } from "../core/types";
import {
	findUserPluginInfo,
	isDefaultPluginUrl,
	normalizePluginInstallUrl,
} from "../core/utils";
import { addLog } from "./logs";

export async function reloadPluginDefinitions(): Promise<void> {
	setBusy(true, "Loading plugin definitions", 10);
	try {
		const [userPluginInfos, corePlugins, installedTools, remoteInfos] = await Promise.all([
			rpc.request.listInstalledUserPluginInfos(),
			rpc.request.listCorePluginNames(),
			rpc.request.listInstalledToolNames(),
			rpc.request.listRemotePluginInfos(),
		]);
		state.installedUserPluginInfos = userPluginInfos;
		state.installedPluginNames = userPluginInfos.map((entry) => entry.name);
		state.corePluginNames = corePlugins;
		state.installedToolNames = installedTools;
		state.remotePluginInfos = remoteInfos;
		state.remotePluginNames = [...new Set(remoteInfos.map((entry) => entry.name))].sort((a, b) =>
			a.localeCompare(b),
		);
		state.installsLoaded = true;
		addLog(`Loaded install sources: remotePlugins=${state.remotePluginNames.length}, userPlugins=${state.installedPluginNames.length}, corePlugins=${state.corePluginNames.length}, installedTools=${state.installedToolNames.length}.`);
		setBusy(false, "Ready", 0);
	} catch (error) {
		addLog(`Failed to load plugin definitions: ${(error as Error).message}`);
		setBusy(false, "Load failed", 0);
	}
}

export async function installPluginDefinition(
	plugin: string,
	gitUrl?: string,
	force = false,
	removeToolAlias = false,
): Promise<void> {
	if (state.busy) {
		return;
	}
	setBusy(true, `${force ? "Updating" : "Installing"} plugin ${plugin}`, 40);
	try {
		const result = await rpc.request.installPluginDefinition({
			plugin,
			gitUrl,
			force,
			...(removeToolAlias ? { removeToolAlias: true } : {}),
		});
		await reloadPluginDefinitions();
		const urlText = gitUrl && gitUrl.trim().length > 0 ? ` (${gitUrl.trim()})` : "";
		addLog(`${plugin}: plugin ${force ? "updated" : "installed"}${urlText}.`);
		if (
			result.stdout.includes("Updated tool_alias") ||
			result.stdout.includes("Removed tool_alias")
		) {
			addLog(`${plugin}: ${result.stdout.split("\n").at(-1) ?? ""}`);
		}
	} catch (error) {
		const actionLabel = force ? "Update" : "Install";
		addLog(`${plugin}: plugin ${force ? "update" : "install"} failed - ${(error as Error).message}`);
		setBusy(false, `${actionLabel} failed: ${plugin}`, 0);
	}
}

export function openInstallPluginDialog(plugin: string): void {
	state.pendingPluginUrlDialog = { mode: "install", pluginName: plugin };
	state.pendingPluginNameValue = "";
	state.pendingPluginUrlValue = "";
}

export function openCustomPluginDialog(): void {
	state.pendingPluginUrlDialog = { mode: "custom-install" };
	state.pendingPluginNameValue = "";
	state.pendingPluginUrlValue = "";
}

export function openEditPluginDialog(plugin: string): void {
	const userInfo = findUserPluginInfo(plugin, state.installedUserPluginInfos);
	if (!userInfo) {
		addLog(`${plugin}: cannot edit URL because user plugin is not installed.`);
		return;
	}
	state.pendingPluginUrlDialog = { mode: "edit", pluginName: plugin };
	state.pendingPluginNameValue = "";
	state.pendingPluginUrlValue = userInfo.url ?? "";
}

export function closePluginUrlDialog(): void {
	state.pendingPluginUrlDialog = null;
	state.pendingPluginNameValue = "";
	state.pendingPluginUrlValue = "";
}

export type PluginInstallPlan =
	| { kind: "invalid"; log: string }
	| { kind: "skip"; log: string }
	| {
			kind: "install";
			plugin: string;
			gitUrl?: string;
			force: boolean;
			removeToolAlias: boolean;
	  };

// Pure decision core of the plugin URL dialog: what to send to mise, or why
// not. "invalid" keeps the dialog open; "skip" closes it without installing.
export function resolvePluginInstallPlan(input: {
	dialog: NonNullable<MainViewState["pendingPluginUrlDialog"]>;
	nameValue: string;
	urlValue: string;
	installedUserPluginInfos: PluginDefinitionInfo[];
	remotePluginInfos: PluginDefinitionInfo[];
}): PluginInstallPlan {
	const rawGitUrl = input.urlValue.trim();
	const gitUrl = normalizePluginInstallUrl(rawGitUrl);

	if (input.dialog.mode === "custom-install") {
		const pluginName = input.nameValue.trim();
		if (pluginName.length === 0) {
			return { kind: "invalid", log: "Custom plugin: plugin name required." };
		}
		if (!/^[A-Za-z0-9._-]+$/.test(pluginName)) {
			return {
				kind: "invalid",
				log: `${pluginName}: plugin name can only contain letters, numbers, dot, underscore, dash.`,
			};
		}
		if (gitUrl.length === 0) {
			return {
				kind: "invalid",
				log: `${pluginName}: Git URL is required for custom plugin install.`,
			};
		}
		return { kind: "install", plugin: pluginName, gitUrl, force: false, removeToolAlias: false };
	}

	if (input.dialog.mode === "edit" && rawGitUrl.length === 0) {
		return { kind: "invalid", log: `${input.dialog.pluginName}: URL is required for edit.` };
	}

	let nextGitUrl: string | undefined = gitUrl.length > 0 ? gitUrl : undefined;
	let removeToolAlias = false;

	if (input.dialog.mode === "edit") {
		const userInfo = findUserPluginInfo(
			input.dialog.pluginName,
			input.installedUserPluginInfos,
		);
		const nextIsDefault = isDefaultPluginUrl(
			input.dialog.pluginName,
			nextGitUrl,
			input.remotePluginInfos,
		);
		if (nextIsDefault) {
			nextGitUrl = undefined;
			removeToolAlias = userInfo?.source === "tool_alias";
		}
		if (!removeToolAlias && (userInfo?.url ?? "").trim() === rawGitUrl) {
			return {
				kind: "skip",
				log: `${input.dialog.pluginName}: URL unchanged; plugin update skipped.`,
			};
		}
	}

	return {
		kind: "install",
		plugin: input.dialog.pluginName,
		gitUrl: nextGitUrl,
		force: input.dialog.mode === "edit",
		removeToolAlias,
	};
}

export async function submitPluginUrlDialog(): Promise<void> {
	if (!state.pendingPluginUrlDialog || state.busy) {
		return;
	}

	const plan = resolvePluginInstallPlan({
		dialog: state.pendingPluginUrlDialog,
		nameValue: state.pendingPluginNameValue,
		urlValue: state.pendingPluginUrlValue,
		installedUserPluginInfos: state.installedUserPluginInfos,
		remotePluginInfos: state.remotePluginInfos,
	});

	if (plan.kind === "invalid") {
		addLog(plan.log);
		return;
	}

	closePluginUrlDialog();
	if (plan.kind === "skip") {
		addLog(plan.log);
		return;
	}
	await installPluginDefinition(plan.plugin, plan.gitUrl, plan.force, plan.removeToolAlias);
}

export async function uninstallPluginDefinition(plugin: string): Promise<void> {
	if (state.busy) {
		return;
	}
	setBusy(true, `Removing plugin ${plugin}`, 40);
	try {
		await rpc.request.uninstallPluginDefinition({ plugin });
		await reloadPluginDefinitions();
		addLog(`${plugin}: plugin removed.`);
	} catch (error) {
		addLog(`${plugin}: plugin remove failed - ${(error as Error).message}`);
		setBusy(false, "Ready", 0);
	}
}

File: Packages/MiseCore/Sources/MiseCore/Features/Installs.swift
/// Plugin definitions (installs tab). Port of src/mainview/features/installs.ts.
public enum PluginInstallPlan: Equatable, Sendable {
    /// Keeps the dialog open.
    case invalid(log: String)
    /// Closes the dialog without installing.
    case skip(log: String)
    case install(plugin: String, gitUrl: String?, force: Bool, removeToolAlias: Bool)
}

extension AppState {
    public func reloadPluginDefinitions() async {
        guard !busy else { return }
        await loadPluginDefinitions()
    }

    private func loadPluginDefinitions() async {
        setBusy(true, Strings.loadingPluginDefinitions)
        installsError = nil
        do {
            async let users = mise.installedUserPluginInfos()
            async let core = mise.corePluginNames()
            async let tools = mise.installedToolNames()
            async let remote = mise.remotePluginInfos()
            let (userInfos, coreNames, toolNames, remoteInfos) = try await (users, core, tools, remote)
            installedUserPluginInfos = userInfos
            installedPluginNames = userInfos.map(\.name)
            corePluginNames = coreNames
            installedToolNames = toolNames
            remotePluginInfos = remoteInfos
            remotePluginNames = Array(Set(remoteInfos.map(\.name))).sorted { $0.localizedCompare($1) == .orderedAscending }
            installsLoaded = true
            addLog("Loaded install sources: remotePlugins=\(remotePluginNames.count), userPlugins=\(installedPluginNames.count), corePlugins=\(corePluginNames.count), installedTools=\(installedToolNames.count).")
            setBusy(false)
        } catch {
            installsError = error.localizedDescription
            addLog("Failed to load plugin definitions: \(error.localizedDescription)")
            setBusy(false, Strings.loadFailed)
        }
    }

    public func checkPluginDefinitionUpdates() async {
        if busy { return }
        setBusy(true, Strings.checkingPluginUpdates)
        pluginUpdatesError = nil
        defer { pluginUpdatesCheckedAt = Self.nowLabel(); setBusy(false) }
        do {
            outdatedPluginNames = try await mise.outdatedPluginDefinitions()
            addLog("Plugin updates: \(outdatedPluginNames.count) available.")
        } catch {
            outdatedPluginNames = []
            pluginUpdatesError = error.localizedDescription
            addLog("Plugin update check failed: \(error.localizedDescription)")
        }
    }

    public func updatePluginDefinition(_ plugin: String) async {
        guard !busy, !updateCheckRunning, pluginUpdatesError == nil, outdatedPluginNames.contains(plugin) else { return }
        setBusy(true, Strings.updatingPluginDefinition(plugin))
        pluginUpdateResult = nil
        var updated = false
        do {
            _ = try await mise.updatePluginDefinition(plugin)
            updated = true
            pluginUpdateResult = Strings.pluginUpdated(plugin)
            // A plugin can change its remote version list; discard all old candidates.
            for index in plugins.indices { plugins[index].status = .idle }
            toolsCheckedAt = nil
            addLog(pluginUpdateResult!)
        } catch {
            pluginUpdateResult = Strings.pluginUpdateFailed(plugin, error.localizedDescription)
            addLog(pluginUpdateResult!)
        }
        setBusy(false)
        if updated { await checkPluginDefinitionUpdates() }
    }

    public func installPluginDefinition(_ plugin: String, gitUrl: String? = nil, force: Bool = false, removeToolAlias: Bool = false) async {
        if busy { return }
        setBusy(true, Strings.installingPlugin(plugin, force: force))
        do {
            let result = try await mise.installPluginDefinition(plugin: plugin, gitUrl: gitUrl, force: force, removeToolAlias: removeToolAlias)
            await loadPluginDefinitions()
            let url = gitUrl?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            addLog("\(plugin): plugin \(force ? "updated" : "installed")\(url.isEmpty ? "" : " (\(url))").")
            if result.stdout.contains("Updated tool_alias") || result.stdout.contains("Removed tool_alias") {
                addLog("\(plugin): \(result.stdout.split(separator: "\n").last ?? "")")
            }
        } catch {
            addLog("\(plugin): plugin \(force ? "update" : "install") failed - \(error.localizedDescription)")
            setBusy(false, Strings.pluginInstallFailed(plugin, force: force))
        }
    }

    public func uninstallPluginDefinition(_ plugin: String) async {
        if busy { return }
        setBusy(true, Strings.removingPlugin(plugin))
        do {
            _ = try await mise.uninstallPluginDefinition(plugin)
            await loadPluginDefinitions()
            addLog("\(plugin): plugin removed.")
        } catch {
            addLog("\(plugin): plugin remove failed - \(error.localizedDescription)")
            setBusy(false)
        }
    }

    // MARK: Plugin URL dialog

    public func openInstallPluginDialog(_ plugin: String) {
        pendingPluginUrlDialog = .install(pluginName: plugin)
        pendingPluginNameValue = ""
        pendingPluginUrlValue = ""
    }

    public func openCustomPluginDialog() {
        pendingPluginUrlDialog = .customInstall
        pendingPluginNameValue = ""
        pendingPluginUrlValue = ""
    }

    public func openEditPluginDialog(_ plugin: String) {
        guard let userInfo = PluginSource.userPluginInfo(plugin, in: installedUserPluginInfos) else {
            addLog("\(plugin): cannot edit URL because user plugin is not installed.")
            return
        }
        pendingPluginUrlDialog = .edit(pluginName: plugin)
        pendingPluginNameValue = ""
        pendingPluginUrlValue = userInfo.url ?? ""
    }

    public func closePluginUrlDialog() {
        pendingPluginUrlDialog = nil
        pendingPluginNameValue = ""
        pendingPluginUrlValue = ""
    }

    /// Pure decision core of the plugin URL dialog: what to send to mise, or why not.
    public static func resolvePluginInstallPlan(
        dialog: PluginUrlDialog, nameValue: String, urlValue: String,
        installedUserPluginInfos: [PluginDefinitionInfo], remotePluginInfos: [PluginDefinitionInfo]
    ) -> PluginInstallPlan {
        let rawGitUrl = urlValue.trimmingCharacters(in: .whitespacesAndNewlines)
        let gitUrl = PluginSource.normalizeInstallURL(rawGitUrl)

        switch dialog {
        case .customInstall:
            let name = nameValue.trimmingCharacters(in: .whitespacesAndNewlines)
            if name.isEmpty { return .invalid(log: "Custom plugin: plugin name required.") }
            if !name.allSatisfy({ $0.isASCII && ($0.isLetter || $0.isNumber || "._-".contains($0)) }) {
                return .invalid(log: "\(name): plugin name can only contain letters, numbers, dot, underscore, dash.")
            }
            if gitUrl.isEmpty { return .invalid(log: "\(name): Git URL is required for custom plugin install.") }
            return .install(plugin: name, gitUrl: gitUrl, force: false, removeToolAlias: false)

        case .install(let pluginName):
            return .install(plugin: pluginName, gitUrl: gitUrl.isEmpty ? nil : gitUrl, force: false, removeToolAlias: false)

        case .edit(let pluginName):
            if rawGitUrl.isEmpty { return .invalid(log: "\(pluginName): URL is required for edit.") }
            let userInfo = PluginSource.userPluginInfo(pluginName, in: installedUserPluginInfos)
            var nextGitUrl: String? = gitUrl.isEmpty ? nil : gitUrl
            var removeToolAlias = false
            if PluginSource.isDefaultURL(pluginName, gitUrl: nextGitUrl, remote: remotePluginInfos) {
                nextGitUrl = nil
                removeToolAlias = userInfo?.source == "tool_alias"
            }
            if !removeToolAlias, (userInfo?.url ?? "").trimmingCharacters(in: .whitespacesAndNewlines) == rawGitUrl {
                return .skip(log: "\(pluginName): URL unchanged; plugin update skipped.")
            }
            return .install(plugin: pluginName, gitUrl: nextGitUrl, force: true, removeToolAlias: removeToolAlias)
        }
    }

    public func submitPluginUrlDialog() async {
        guard let dialog = pendingPluginUrlDialog, !busy else { return }
        let plan = Self.resolvePluginInstallPlan(
            dialog: dialog, nameValue: pendingPluginNameValue, urlValue: pendingPluginUrlValue,
            installedUserPluginInfos: installedUserPluginInfos, remotePluginInfos: remotePluginInfos)
        switch plan {
        case .invalid(let log):
            addLog(log)
        case .skip(let log):
            closePluginUrlDialog()
            addLog(log)
        case .install(let plugin, let gitUrl, let force, let removeToolAlias):
            closePluginUrlDialog()
            await installPluginDefinition(plugin, gitUrl: gitUrl, force: force, removeToolAlias: removeToolAlias)
        }
    }
}
